google-cloud-functions/google_api_auth.py

```diff
@@ -0,0 +1,152 @@
+            self.host = "https://{0}".format(self.host)
+        elif self.host[:8] == "https://":
+            self.host = "{0}".format(self.host)
+        else:
+            self.host = "https://{0}".format(self.host)
+        self.refresh_token = ""
+        self.code = config.get("code")
+        self.token_url = 'https://accounts.google.com/o/oauth2/token'
+        if not config.get("redirect_url"):
+            self.redirect_url = DEFAULT_REDIRECT_URL
+        else:
+            self.redirect_url = config.get("redirect_url")
+
+    def convert_ts_epoch(self, ts):
+        datetime_object = datetime.strptime(ctime(ts), "%a %b %d %H:%M:%S %Y")
+        return datetime_object.timestamp()
+
+    def generate_token(self, REFRESH_TOKEN_FLAG):
+        try:
+            token_resp = acquire_token(self, REFRESH_TOKEN_FLAG)
+            ts_now = time()
+            token_resp['expiresOn'] = (ts_now + token_resp['expires_in']) if token_resp.get("expires_in") else None
+            token_resp['accessToken'] = token_resp.get("access_token")
+            token_resp.pop("access_token")
+            return token_resp
+        except Exception as err:
+            logger.error("{0}".format(err))
+            raise ConnectorError("{0}".format(err))
+
+    def validate_token(self, connector_config, connector_info):
+        try:
+            ts_now = time()
+            if not connector_config.get('accessToken'):
+                logger.error('Error occurred while connecting server: Unauthorized')
+                raise ConnectorError('Error occurred while connecting server: Unauthorized')
+            expires = connector_config['expiresOn']
+            expires_ts = self.convert_ts_epoch(expires)
+            if ts_now > float(expires_ts):
+                REFRESH_TOKEN_FLAG = True
+                logger.info("Token expired at {0}".format(expires))
+                self.refresh_token = connector_config['refresh_token']
+                token_resp = self.generate_token(REFRESH_TOKEN_FLAG)
+                connector_config['accessToken'] = token_resp['accessToken']
+                connector_config['expiresOn'] = token_resp['expiresOn']
+                if token_resp.get('refresh_token'):
+                    connector_config['refresh_token'] = token_resp.get('refresh_token')
+                update_connnector_config(connector_info['connector_name'], connector_info['connector_version'],
+                                         connector_config,
+                                         connector_config['config_id'])
+
+                return "Bearer {0}".format(connector_config.get('accessToken'))
+            else:
+                logger.info("Token is valid till {0}".format(expires))
+                return "Bearer {0}".format(connector_config.get('accessToken'))
+        except Exception as err:
+            logger.error("{0}".format(str(err)))
+            raise ConnectorError("{0}".format(str(err)))
+
+
+def acquire_token(self, REFRESH_TOKEN_FLAG):
```

Comment:
this function parameter name can be changed as there is constant declared in this file with same name